authz_workspace/tag_based_pass_access_policy.py

```diff
@@ -0,0 +1,492 @@
+    last_cycle_seconds = cycles_since_hour * tick_rate
+    next_cycle_seconds = last_cycle_seconds + tick_rate
+    next_cycle_at = now.replace(minute=0, second=0, microsecond=0) + timedelta(
+        seconds=next_cycle_seconds
+    )
+    next_tick = loop.time() + (next_cycle_at - now).total_seconds()
+
+    while True:
+        await asyncio.sleep(max(0, next_tick - loop.time()))
+        now = datetime.now()
+        logger.debug(f"Tag Sync Event Loop Tick {now}")
+        _sync_tags_run_tasks(loop, now)
+        next_tick += tick_rate
+
+
+def _sync_tags_start_loop(loop: asyncio.AbstractEventLoop):
+    asyncio.set_event_loop(loop)
+    loop.create_task(_sync_tags_scheduler())
+    loop.run_forever()
+
+
+def _sync_tags_exc_handler(loop, context):
+    task = context.get("task") or context.get("future")
+    if task is None:
+        msg = context.get("message")
+        logger.debug(f"Exception handler invoked, but there is no task: {msg}")
+
+    e = task.exception()
+    e_type = e.__class__.__name__
+    coro_name = task.get_coro().__qualname__
+    logger.exception(f"Caught '{e_type}' exception from '{coro_name}': {e}")
+
+
+sync_tags_loop = asyncio.new_event_loop()
+sync_tags_loop.set_exception_handler(_sync_tags_exc_handler)
+sync_tags_thread = threading.Thread(
+    target=_sync_tags_start_loop, args=(sync_tags_loop,), daemon=True
+)
+sync_tags_thread.start()
+
+
+def get_group_users(groupname):
+    try:
+        usernames = group_record_cache[groupname]
+    except KeyError:
+        logger.debug("%s: Cache miss on group record", groupname)
+        usernames = grp.getgrnam(groupname).gr_mem
+        group_record_cache[groupname] = usernames
+    else:
+        logger.debug("%s: Cache hit on group record", groupname)
+
+    return usernames
+
+
+def load_tags_file(tags_db):
+    if not tags_db.is_file():
+        raise FileNotFoundError(f"sqlite DB not found: {tags_db}")
+
+    conn = sqlite3.connect(tags_db)
+    cursor = conn.cursor()
+    cursor.execute("SELECT obj_blob FROM store WHERE obj_name = ?", ("tags_state",))
+    row = cursor.fetchone()
+    loaded_tags = pickle.loads(row[0]) if row else None
+    if not loaded_tags:
+        raise ValueError(f"Could not load tags state from DB: {tags_db}")
+
+    return loaded_tags
+
+
+def dump_tags_file(tags_db, loaded_tags):
+    conn = sqlite3.connect(tags_db)
+    cursor = conn.cursor()
+    cursor.execute(
+        """
+CREATE TABLE IF NOT EXISTS store (
+    obj_name TEXT PRIMARY KEY,
+    obj_blob BLOB
+)
+"""
+    )
+    loaded_tags_blob = sqlite3.Binary(
+        pickle.dumps(loaded_tags, protocol=pickle.HIGHEST_PROTOCOL)
+    )
+    cursor.execute(
+        "REPLACE INTO store (obj_name, obj_blob) VALUES (?, ?)",
+        ("tags_state", loaded_tags_blob),
+    )
+    conn.commit()
+
+
+class TagBasedPASSAccessPolicy(TagBasedAccessPolicy):
+    def __init__(self, url, tags_db, sync_proposals={}, *args, **kwargs):
```

Comment:
Avoid mutable defaults. (This can be `None`, and the in the method body, `sync_proposals = sync_proposals or {}`.)